USHIFT-5765: common_versions.sh script generator

vanhalenar opened the pull request:
This PR is the first step: it introduces a script to generate the variables for `common_versions.sh`. Next steps are up for discussion (making it part of rebase).

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+        return f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/microshift/ocp/latest-4.{minor}/el9/os"
+
+
+def get_beta_repo(minor):
```
I would add an extra argument to let know if the function should use `[Y, Y-1, Y-2]` loop or should only try the `Y` and fail (For the previous minor version it should always exist).

vanhalenar replied:
Done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+"""
```
```suggestion
#!/usr/bin/env python

"""
```

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+print(f"export FAKE_NEXT_MINOR_VERSION={fake_next_minor_version}")
+print("")
+
+print(f"CURRENT_RELEASE_REPO={current_release_repo}")
```
common_versions.sh must generate arch agnostic url, we cannot commit this because it has hardcoded arch:
```
CURRENT_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/aarch64/microshift/ocp/latest-4.19/el9/os"
```

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+
+    ``get_candidate_repo`` returns the URL of the EC/RC repository for the specified minor version.
+    """
+    if dev_preview:
+        return f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/microshift/ocp-dev-preview/latest-4.{minor}/el9/os"
+    else:
+        return f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/microshift/ocp/latest-4.{minor}/el9/os"
```
This could be
```suggestion
    return f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/microshift/ocp{'-dev-preview' if dev_preview else ''}/latest-4.{minor}/el9/os"

```

vanhalenar replied:
Done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+            return False
+
+
+def dnf_repo_is_enabled(repo_name):
+    """
+    :param repo_name: the name of the repository to check
+
+    ``dnf_repo_is_enabled`` checks if the repository is enabled on this system.
+    """
+    try:
+        result = subprocess.run(
+            ["dnf", "repolist"],
+            stdout=subprocess.PIPE,
+            stderr=subprocess.DEVNULL,
+            text=True,
+            check=True
+        )
+        return repo_name in result.stdout
+    except subprocess.CalledProcessError:
+        return False
```
This is unused

vanhalenar replied:
Fixed

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+        return False
+
+
+def repo_exists(repo_url):
```
I think function name should hint that it's checking http/mirror repos

vanhalenar replied:
Done

vanhalenar replied:
Done, renamed to `mirror_exists`

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+        return True
+
+
+def get_sub_repo(minor):
```
I think we can use full name and hint on the functionality, e.g. `get_subscription_repo_name_if_exists`

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+    """
+    try:
+        repo = f"rhocp-4.{minor}-for-rhel-9-{ARCH}-rpms"
+        _ = subprocess.run(
+                ['sudo', 'dnf', 'repoquery', 'microshift', '--quiet',
+                 '--queryformat', '%{version}-%{release}', '--repo', repo],
+                capture_output=True,
+                text=True,
+                check=True
+            )
```
Why not use `provides_pkg()`?

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+    in that order. If none of these repositories are available, returns empty
+    double quotes (to be used as empty string in bash).
+    """
+    repo = get_sub_repo(minor)
+    if repo is not None and provides_pkg(repo, "microshift"):
```
I think it's a bit cyclic.
In first call we're checking if rhocp for `4.$minor` contains microshift, then use `provides_pkg(repo, "microshift")`. after checking if `repo is not None` we already know that it's there.

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+# The current release repository comes from the 'rhocp' stream for release
+# branches, or the OpenShift mirror if only a RC or EC is available. It can
+# be empty, if no candidate for the current minor has been built yet.
+current_release_repo = get_release_repo(minor_version)
```
I would rename `get_release_repo` to include the fact it's looking for microshift. 
Something like `get_microshift_repo` or `get_published_microshift_repo`

vanhalenar replied:
Done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+# The beta repository, containing dependencies, should point to the
+# OpenShift mirror URL. If the repository for current minor is not
+# available yet, it should point to an older release.
+rhocp_minor_y_beta = get_beta_repo(minor_version)
```
I would rename `get_beta_repo` to something like `get_beta_dependencies_repo`

vanhalenar replied:
Done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+
+# The 'rhocp_minor_y' variable should be the minor version number, if the
+# current release is available through the 'rhocp' stream, otherwise empty.
+rhocp_minor_y = minor_version if get_sub_repo(minor_version) is not None else '""'
```
I'd rename `get_sub_repo` to mentions it is for dependencies, like `get_subscription_dependencies_repo`

vanhalenar replied:
`get_sub_repo` is not for dependencies though, it's for getting the subscription repository. I renamed it to `get_subscription_repo_name_if_exists` as per earlier comment.

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+    is a URL or not. if neither, returns empty double quotes.
+    """
+    if repo.startswith("\"rhocp"):
+        return f'"$(get_vrel_from_rhsm {repo})"'
+    elif repo.startswith("\"https"):
+        return f'"$(get_vrel_from_beta {repo})"'
```
Ideally, we want them to still reference variables, not urls.

That is:
```
PREVIOUS_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/$(uname -m)/microshift/ocp/latest-4.19/el9/os"
PREVIOUS_RELEASE_VERSION="$(get_vrel_from_beta "${PREVIOUS_RELEASE_REPO}")"
```
instead of
```sh
PREVIOUS_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/aarch64/microshift/ocp/latest-4.19/el9/os"
PREVIOUS_RELEASE_VERSION="$(get_vrel_from_beta "https://mirror.openshift.com/pub/openshift-v4/aarch64/microshift/ocp/latest-4.19/el9/os")"
```

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+    for i in range(minor, minor-3, -1):
+        url = f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/dependencies/rpms/4.{i}-el9-beta"
+        if repo_exists(url) and provides_pkg(url, "cri-o"):
+            return f'"{url}"'
```
I think we could add quotes as last step, when rendering the file and otherwise keep it clean
```suggestion
            return url
```

vanhalenar replied:
Done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+                     '--repo', repo],
+                    capture_output=True,
+                    check=True
+            )
```
Done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+
+cncf_sonobuoy_version = "v0.57.3"
+
+print(f"export MINOR_VERSION={minor_version}")
```
Done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+print(f"export MINOR_VERSION={minor_version}")
+print(f"export PREVIOUS_MINOR_VERSION={previous_minor_version}")
+print(f"export YMINUS2_MINOR_VERSION={yminus2_minor_version}")
+print(f"export FAKE_NEXT_MINOR_VERSION={fake_next_minor_version}")
```
Done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,253 @@
+# The 'rhocp_minor_y2' should always be the y-2 minor version number.
+rhocp_minor_y2 = yminus2_minor_version
+
+cncf_sonobuoy_version = "v0.57.3"
```
Done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,230 @@
+
+    if repo.startswith("https"):
+        temp = f"this,{repo}"
+        args += ['--disablerepo', '*', '--repofrompath', temp]
```
done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,230 @@
+        args += ['--repo', repo]
+
+    try:
+        subprocess.run(args, stdout=sys.stderr, check=True)
```
done

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,230 @@
+
+# The 'rhocp_minor_y1' should always be the y-1 minor version number.
+# The repository for y-1 release should always exist.
+rhocp_minor_y1 = previous_minor_version
```
removed

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,230 @@
+minor_version = args.minor
+previous_minor_version = minor_version - 1
+yminus2_minor_version = minor_version - 2
+fake_next_minor_version = minor_version + 1
```
removed

vanhalenar commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,230 @@
+    :param minor: the minor version, e.g. 19 for 4.19
+    :param dev_preview: if True, returns the engineering candidate repo, otherwise returns the release candidate repo
+
+    ``get_candidate_repo_url`` returns the URL of the EC/RC repository for the specified minor version.
```
Changed. Disclaimer, I was basing the comments on `qemu_guest_agent.py`.

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+# The 'rhocp_minor_y2' should always be the y-2 minor version number.
+rhocp_minor_y2 = yminus2_minor_version
+
+output = f"""#!/bin/bash
+set -euo pipefail
```
Let's add a comment that this file auto-generated using this script?
```suggestion
output = f"""#!/bin/bash

# Following file is auto-generated using generate_common_versions.py.
# It should not be edited manually.

set -euo pipefail
```

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+    else:
+        args += ['--repo', repo]
+
+    try:
```
```suggestion
    try:
         logging.info(f"Running command: {' '.join(args)}")
```

vanhalenar replied:
done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+        return ec
+
+    logging.info(f"No repository found for 4.{minor}")
+    return ""
```
In other functions on "failure" we return `None`. Want to unify?

vanhalenar replied:
done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+    elif repo.startswith("https"):
+        return f'$(get_vrel_from_beta "${{{var_name}}}")'
+    else:
+        return ""
```
```suggestion
        logging.warning("Received unexpected {repo=}")
        return ""
```

vanhalenar replied:
nut sure we want this, it's valid for repo to be empty string here

vanhalenar replied:
I added a case for the empty string, which is valid, and included the warning under `else`

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+    for i in range(minor, minor-steps_back-1, -1):
+        url = f"https://mirror.openshift.com/pub/openshift-v4/{ARCH}/dependencies/rpms/4.{i}-el9-beta"
+        if mirror_exists(url) and repo_provides_pkg(url, "cri-o"):
+            logging.info(f"Beta repository found for 4.{i}")
```
```suggestion
            logging.info(f"Beta dependencies repository found for 4.{i}")
```

vanhalenar replied:
done

pmtk commented on test/bin/generate_common_versions.py:
```diff
@@ -0,0 +1,322 @@
+    Returns:
+        str or None: The URL of a valid beta repository, or None if none are found.
+    """
+    logging.info(f"Getting beta repository for 4.{minor}, max. {steps_back} previous minors")
```
```suggestion
    logging.info(f"Getting beta dependencies repository for 4.{minor}, max. {steps_back} previous minors")
```

pmtk reviewed (commented):
Very nice.
Only minor comments mostly regarding logs :)

pmtk commented:
/retest
/lgtm

ggiguash commented on test/bin/common_versions.sh:
```diff
@@ -39,66 +42,37 @@ get_vrel_from_rhsm() {
 export YMINUS2_MINOR_VERSION=$(( "${MINOR_VERSION}" - 2 ))
 export FAKE_NEXT_MINOR_VERSION=$(( "${MINOR_VERSION}" + 1 ))
 
-# For a main branch, the current release repository usually comes from
```
If we generate this file, we should keep the comments too - if not in the generated file, than in the generator code.

These comments contain valuable information.

vanhalenar replied:
I've reworded the comments a bit and included them in the generator code, is there any specific information that is missing?

agullon commented:
Not sure about how or who should execute this new script, but I suggest you add a bash script to run the python one.
Move this new python script into `pyutils` (maybe you can reuse something from `common.py`) and use a similar bash script like [build_bootc_images.sh](https://github.com/openshift/microshift/blob/main/test/bin/build_bootc_images.sh) to execute the python script.

agullon commented:
I also suggest splitting the new python script in 2 files: one with the code and another one with the content in `output` var. You can name this last one `common_versions.sh.template` or something similar. I think this is a good practice because this new _template_ is quite big, and it's easier to review changes in the future.

pmtk commented:
> Not sure about how or who should execute this new script, but I suggest you add a bash script to run the python one. Move this new python script into `pyutils` (maybe you can reuse something from `common.py`) and use a similar bash script like [build_bootc_images.sh](https://github.com/openshift/microshift/blob/main/test/bin/build_bootc_images.sh) to execute the python script.

Maybe we should have one common venv for all the scripts? Then it would be just matter of running `./script.py`

agullon commented on test/bin/common_versions.sh:
```diff
@@ -135,12 +112,3 @@ export BREW_EC_RELEASE_VERSION
 
 LATEST_RELEASE_TYPE="ec"
 export LATEST_RELEASE_TYPE
-
-BREW_LREL_RELEASE_VERSION="${BREW_EC_RELEASE_VERSION}"
-export BREW_LREL_RELEASE_VERSION
-
-# Branch and commit for the openshift-tests-private repository
-OPENSHIFT_TESTS_PRIVATE_REPO_BRANCH="release-4.${MINOR_VERSION}"
-OPENSHIFT_TESTS_PRIVATE_REPO_COMMIT="61613d96c91db7b2907c24dd257075d3f2201991"
-export OPENSHIFT_TESTS_PRIVATE_REPO_BRANCH
-export OPENSHIFT_TESTS_PRIVATE_REPO_COMMIT
```
please do not remove this lines added last week

pmtk commented on test/bin/generate_common_versions.sh:
Why do we need this wrapper? `generate_common_versions.py` is not using any venv, so this is (to me) unnecessary file duplication

pmtk commented on test/bin/common_versions.sh:
```diff
@@ -130,18 +103,17 @@ export CNCF_SYSTEMD_LOGS_VERSION=v0.4
-BREW_Y2_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${YMINUS2_MINOR_VERSION}-zstream/${UNAME_M}/")"
-BREW_RC_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${MINOR_VERSION}-rc/${UNAME_M}/")"
-BREW_EC_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${MINOR_VERSION}-ec/${UNAME_M}/")"
+BREW_Y0_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${MINOR_VERSION}-zstream/$(uname -m)/")"
+BREW_Y1_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${PREVIOUS_MINOR_VERSION}-zstream/$(uname -m)/")"
+BREW_Y2_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${YMINUS2_MINOR_VERSION}-zstream/$(uname -m)/")"
+BREW_RC_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${MINOR_VERSION}-rc/$(uname -m)/")"
+BREW_EC_RELEASE_VERSION="$(get_vrel_from_rpm "${BREW_RPM_SOURCE}/4.${MINOR_VERSION}-ec/$(uname -m)/")"
```
We could keep the `${UNAME_M}` to minimize amount of spawned subprocesses

agullon commented on test/bin/common_versions.sh:
```diff
@@ -59,68 +62,38 @@ get_vrel_from_rpm() {
-# official 'rhocp' stream.
-CURRENT_RELEASE_REPO="" # "https://mirror.openshift.com/pub/openshift-v4/$(uname -m)/microshift/ocp/latest-4.21/el9/os"
-CURRENT_RELEASE_VERSION="" # "$(get_vrel_from_beta "${CURRENT_RELEASE_REPO}")"
+CURRENT_RELEASE_REPO="https://mirror.openshift.com/pub/openshift-v4/$(uname -m)/microshift/ocp-dev-preview/latest-4.21/el9/os"
+CURRENT_RELEASE_VERSION="$(get_vrel_from_beta "${CURRENT_RELEASE_REPO}")"
```
afaik these 2 lines must be commented because

vanhalenar commented:
Regarding the comments from `common_versions.sh` that I removed, they're still in `generate_common_versions.py`, just a bit reworded to reflect the changes. Should I include them in the generated `common_versions.sh` or not? I thought that if the script is being auto-generated, it makes more sense to not have comments there.

agullon commented:
> Regarding the comments from `common_versions.sh` that I removed, they're still in `generate_common_versions.py`, just a bit reworded to reflect the changes. Should I include them in the generated `common_versions.sh` or not? I thought that if the script is being auto-generated, it makes more sense to not have comments there.

From the point of view of someone reading the code to understand how it's working, I think there's more value having the comments in the `common_versions.sh`. The reason is, for example: if I'm trying to understand how `scenarios.sh` works I need to understand how the vars in `common_versions.sh` are defined, so it'll be easier to have the comments right there.

vanhalenar commented:
> > Regarding the comments from `common_versions.sh` that I removed, they're still in `generate_common_versions.py`, just a bit reworded to reflect the changes. Should I include them in the generated `common_versions.sh` or not? I thought that if the script is being auto-generated, it makes more sense to not have comments there.
> 
> From the point of view of someone reading the code to understand how it's working, I think there's more value having the comments in the `common_versions.sh`. The reason is, for example: if I'm trying to understand how `scenarios.sh` works I need to understand how the vars in `common_versions.sh` are defined, so it'll be easier to have the comments right there.

Alright, I've added back comments to `common_versions.sh` itself.

pmtk commented:
/verified by ci
